src/osd/PeeringState.cc

```diff
@@ -1594,36 +1594,68 @@ void PeeringState::calc_replicated_acting(
   }
 }
 
-bool PeeringState::recoverable_and_ge_min_size(const vector<int> &want) const
-{
-  unsigned num_want_acting = 0;
-  set<pg_shard_t> have;
-  for (int i = 0; i < (int)want.size(); ++i) {
-    if (want[i] != CRUSH_ITEM_NONE) {
-      ++num_want_acting;
-      have.insert(
-        pg_shard_t(
-          want[i],
-          pool.info.is_erasure() ? shard_id_t(i) : shard_id_t::NO_SHARD));
-    }
-  }
-  // We go incomplete if below min_size for ec_pools since backfill
-  // does not currently maintain rollbackability
-  // Otherwise, we will go "peered", but not "active"
-  if (num_want_acting < pool.info.min_size &&
-      (pool.info.is_erasure() ||
-       !cct->_conf->osd_allow_recovery_below_min_size)) {
-    psdout(10) << __func__ << " failed, below min size" << dendl;
-    return false;
-  }
+bool PeeringState::recoverable(const vector<int> &want) const
+{
+
+  if (HAVE_FEATURE(get_osdmap()->get_up_osd_features(), SERVER_OCTOPUS)) {
+    unsigned num_want_acting = 0;
+    set<pg_shard_t> have;
+    for (int i = 0; i < (int)want.size(); ++i) {
+      if (want[i] != CRUSH_ITEM_NONE) {
+	++num_want_acting;
+	have.insert(
+	  pg_shard_t(
+	    want[i],
+	    pool.info.is_erasure() ? shard_id_t(i) : shard_id_t::NO_SHARD));
+      }
+    }
 
-  /* Check whether we have enough acting shards to later perform recovery */
-  if (!missing_loc.get_recoverable_predicate()(have)) {
-    psdout(10) << __func__ << " failed, not recoverable" << dendl;
-    return false;
-  }
+    /* Check whether we have enough acting shards to later perform recovery */
+    if (num_want_acting < pool.info.min_size) {
```

Comment:
shall we still check `missing_loc.get_recoverable_predicate()(have)` even if `num_want_acting >= pool.info.min_size` ?

also, i'd suggest consolidate these checks to improve the readability:
```c++
if (num_want_acting < pool.info.min_size) {
  const bool recovery_ec_pg_under_min_size =
    HAVE_FEATURE(get_osdmap()->get_up_osd_features(), SERVER_OCTOPUS);
  if (pool.info.is_erasure() && !recovery_ec_pg_under_min_size) {
    psdout(10) << __func__ << " failed, ec recovery below min size not supported pre-preoctopus" << dendl;
    return false;
  } else if (!cct->_conf->osd_allow_recovery_below_min_size) {
    psdout(10) << __func__ << " failed, recovery below min size not enabled" << dendl;
    return false;
  }
}
if (missing_loc.get_recoverable_predicate()(have)) {
  return true;
} else {
  psdout(10) << __func__ << " failed, not recoverable " << dendl;
  return false;
}
```

Reply:
thank you, @tchaikov . 

> shall we still check missing_loc.get_recoverable_predicate()(have) even if num_want_acting >= pool.info.min_size ?

yes, we should check it. i wonder if user set wrong min size on ec pool. ec pool is not recoverable even `num_want_acting >= pool.info.min` ( k=3 m=2 min_size=2 num_want_acting=2)